### ISSUE TITLE
mkdir /.talos: permission denied

### ISSUE THREAD

**Issue by darox** [siderolabs/talos-backup#27]
I am seeing the following issue when running the backup tool as Kubernetes Cronjob:

```
2024/12/26 10:25:14 failed to get talosconfig: mkdir /.talos: permission denied
```

Why does it need to create this folder and why does it fail?

I'm using the official example in this repo.

**Comment by darox** [comment]
I got it working by mounting a `TALOSCONFIG` file with the specific permissions:

```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: talos-backup
  namespace: talos
spec:
  schedule: "0 */12 * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: talos-backup
            image: ghcr.io/siderolabs/talos-backup:v0.1.0-beta.2-1-g9ccc125
            workingDir: /tmp
            imagePullPolicy: IfNotPresent
            env:
              - name: AWS_ACCESS_KEY_ID
                valueFrom:
                  secretKeyRef:
                    name: talos-backup-secrets
                    key: AWS_ACCESS_KEY_ID
              - name: AWS_SECRET_ACCESS_KEY
                valueFrom:
                  secretKeyRef:
                    name: talos-backup-secrets
                    key: AWS_SECRET_ACCESS_KEY
              - name: CUSTOM_S3_ENDPOINT
                value: "https://x.r2.cloudflarestorage.com"
              - name: BUCKET
                value: "homelab"
              - name: CLUSTER_NAME
                value: "x"
              - name: S3_PREFIX
                value: "x"
              - name: USE_PATH_STYLE
                value: "false"
              - name: AGE_X25519_PUBLIC_KEY
                value: x
              - name: TALOSCONFIG
                value: /tmp/talosconfig
              - name: AWS_REGION
                value: us-west-2
            securityContext:
              runAsUser: 1
              runAsGroup: 1
              allowPrivilegeEscalation: true
              runAsNonRoot: false
              capabilities:
                drop:
                  - ALL
              seccompProfile:
                type: RuntimeDefault
            command:
            - /talos-backup
            volumeMounts:
            - mountPath: /tmp
              name: tmp
            - mountPath: /var/run/secrets/talos.dev
              name: talos-secrets
            - mountPath: /tmp/talosconfig
              name: talosconfig
              subPath: talosconfig
          restartPolicy: OnFailure
          volumes:
          - emptyDir: {}
            name: tmp
          - name: talos-secrets
            secret:
              secretName: talos-backup-secrets
          - name: talosconfig
            secret:
              secretName: talosconfig
```

**Comment by FlorisFeddema** [comment]
I had the same issue, seems like the talosconfig is written to the .talos folder. I added that folder as an empty dir as well and have it working since today. (see: https://github.com/FlorisFeddema/homelab/blob/main/products/storage/talos-backup/templates/cronjob.yaml)

**Comment by darox** [comment]
> I had the same issue, seems like the talosconfig is written to the .talos folder. I added that folder as an empty dir as well and have it working since today. (see: https://github.com/FlorisFeddema/homelab/blob/main/products/storage/talos-backup/templates/cronjob.yaml)

Intersting. I tried the same with:

```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: {{ .Release.Name }}-talos-backup-etcd
  namespace: {{ .Release.Namespace }}
  labels:
    app.kubernetes.io/name: {{ .Release.Name }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    app.kubernetes.io/managed-by: Helm
spec:
  schedule: {{ .Values.schedule }}
  jobTemplate:
    spec:
      template:
        spec:
          tolerations:
            - key: node-role.kubernetes.io/control-plane
              operator: Exists
              effect: NoSchedule
          nodeSelector:
            node-role.kubernetes.io/control-plane: ""
          containers:
          - name: talos-backup-etcd
            image: {{ .Values.image }}
            workingDir: /tmp
            imagePullPolicy: IfNotPresent
            env:
              - name: AWS_ACCESS_KEY_ID
                valueFrom:
                  secretKeyRef:
                    name: {{ .Values.s3.secretName }}
                    key: AWS_ACCESS_KEY_ID
              - name: AWS_SECRET_ACCESS_KEY
                valueFrom:
                  secretKeyRef:
                    name: talos-backup-secrets
                    key: AWS_SECRET_ACCESS_KEY
              # secret because some S3 providers have things like account ID in the endpoint
              - name: CUSTOM_S3_ENDPOINT
                valueFrom: 
                  secretKeyRef:
                    name: {{ .Values.s3.secretName }}
                    key: CUSTOM_S3_ENDPOINT
              - name: BUCKET
                value: {{ .Values.s3.bucket }}
              - name: CLUSTER_NAME
                value: {{ .Values.clusterName }}
              - name: S3_PREFIX
                value: {{ .Values.s3.prefix }}
              - name: USE_PATH_STYLE
                value: {{ .Values.s3.usePathStyle | quote }}
              - name: AGE_X25519_PUBLIC_KEY
                value: {{ .Values.age.x25519PublicKey }}
              - name: AWS_REGION
                value: {{ .Values.s3.region }}
            securityContext:
              runAsUser: 1000
              runAsGroup: 1000
              allowPrivilegeEscalation: false
              runAsNonRoot: true
              capabilities:
                drop:
                  - ALL
              seccompProfile:
                type: RuntimeDefault
            command:
            - /talos-backup
            volumeMounts:
            - mountPath: /tmp
              name: tmp
            - mountPath: /var/run/secrets/talos.dev
              name: s3-secrets
            - mountPath: "/.talos"
              name: talos
          restartPolicy: {{ .Values.restartPolicy }}
          volumes:
          - emptyDir: {}
            name: tmp
          - emptyDir: {}
            name: talos
          - name: s3-secrets
            secret:
              secretName: {{ .Values.s3.secretName }}
```

But this leads to a:

```
2024/12/31 11:03:41 failed to create talos client: failed to determine endpoints
```

**Comment by darox** [comment]
The secret that gets created has the following setting:

```
context: default
contexts:
    default:
        endpoints:
            - talos.default
```

I think it fails because of `talos.default`

**Comment by FlorisFeddema** [comment]
I had the same error `failed to create talos client: failed to determine endpoints`.

I did not use the correct namespace as shown in the talos config (https://www.talos.dev/v1.9/advanced/talos-api-access-from-k8s/#enabling-the-feature). The feature should be enabled on the nodes that talos-backup runs on. And I believe the feature can only be enabled on control plane nodes.

After I changed the namespace in the talos config, I had to recreate the secret + service account to make them work correctly.
